streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java

```diff
@@ -571,68 +615,155 @@ public GroupAssignment assign(final Cluster metadata, final GroupSubscription gr
+            final ClientState state = clientMetadata.state;
 
-            int consumerTaskIndex = 0;
+            final Map<String, List<TaskId>> interleavedActive =
+                interleaveTasksByGroupId(state.activeTasks(), clientMetadata.consumers);
 
-            for (final String consumer : consumers) {
-                final Map<TaskId, Set<TopicPartition>> standby = new HashMap<>();
-                final List<AssignedPartition> assignedPartitions = new ArrayList<>();
+            addClientAssignments(
+                assignment,
+                clientMetadata,
+                partitionsForTask,
+                partitionsByHostState,
+                allOwnedPartitions,
+                interleavedActive,
+                futureConsumers,
+                minUserMetadataVersion);
+        }
 
-                final List<TaskId> assignedActiveList = interleavedActive.get(consumerTaskIndex);
+        return assignment;
+    }
 
-                for (final TaskId taskId : assignedActiveList) {
-                    for (final TopicPartition partition : partitionsForTask.get(taskId)) {
+    private void addClientAssignments(final Map<String, Assignment> assignment,
+                                      final ClientMetadata clientMetadata,
+                                      final Map<TaskId, Set<TopicPartition>> partitionsForTask,
+                                      final Map<HostInfo, Set<TopicPartition>> partitionsByHostState,
+                                      final Set<TopicPartition> allOwnedPartitions,
+                                      final Map<String, List<TaskId>> activeTaskAssignments,
+                                      final Set<String> futureConsumers,
+                                      final int minUserMetadataVersion) {
+        final ClientState clientState = clientMetadata.state;
+        final Map<String, List<TaskId>> interleavedStandby =
+            interleaveTasksByGroupId(clientState.standbyTasks(), clientMetadata.consumers);
+
+        for (final String consumer : clientMetadata.consumers) {
+            final Map<TaskId, Set<TopicPartition>> standby = new HashMap<>();
+            final List<AssignedPartition> assignedPartitions = new ArrayList<>();
+
+            final List<TaskId> assignedActiveList = activeTaskAssignments.get(consumer);
+
+            for (final TaskId taskId : assignedActiveList) {
+                for (final TopicPartition partition : partitionsForTask.get(taskId)) {
+                    final String oldOwner = clientState.ownedPartitions().get(partition);
+                    final boolean newPartitionForConsumer = oldOwner == null || !oldOwner.equals(consumer);
+
+                    // If the partition is new to this consumer but is still owned by another, remove from the assignment
+                    // until it has been revoked and can safely be reassigned according the COOPERATIVE protocol
```

Comment:
Not sure I follow -- if `clientState.ownedPartitions().get(partition)` is not null then `newPartitionForConsumer` is likely false and the condition is as well (I agree, regardless of `allOwnedPartitions`. But consider the other case, clientState.ownedPartitions().get(partition)` IS null and `newPartitionForConsumer` is true -- we want to make sure this partition was actually claimed by someone else before removing it from the assignment (since it could be completely new)
Does that make sense? Or did I misinterpret your point here

Reply:
My bad, mis read the lines :)